### ISSUE TITLE
Are the old bug tickets from avahi.org still available?

### ISSUE THREAD

**Issue by mspncp** [lathiat/avahi#187]
Hi Trent,

i am currently seeking for more detailed information about the following two bugfixes. (Not because of a bug in avahi, but because of a similar bug in our own software.) 

```
commit 9acc5a79b95ddb136ac53c9c746c4585ac347cb2
Author: Lennart Poettering <lennart@poettering.net>
Date:   Sat Apr 2 03:11:59 2011 +0200

    iface: look for both IFA_ADDRESS and IFA_LOCAL

    http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=617948
    http://avahi.org/ticket/333
```

```
commit 8b3f8e4a63f09d9c9f86dfaddb0681cc612a79eb
Author: Trent Lloyd <trent@lloyd.id.au>
Date:   Mon Dec 1 15:04:33 2014 +0800

    Fix reversed IFA_LOCAL and IFA_ADDRESS checks. (Bug: #355)
```

The commits refer to ticket numbers \#333 and \#355, respectively, which seem to belong to an old ticket system available at http://avahi.org/ticket/xxx. Are these tickets still available online and if yes, where can I find them?

Thanks in advance,
Matthias

**Comment by mspncp** [comment]
I _did_ find the following links in a mail archive. But they provide only ticket status updates. No comments or information about the rationale behind the changes. Or is there any more information?

- http://mail.0pointer.net/pipermail/avahi-tickets/2011-March/001489.html
- http://mail.0pointer.net/pipermail/avahi-tickets/2011-March/001494.html
- http://mail.0pointer.net/pipermail/avahi-tickets/2011-April/001495.html
- http://mail.0pointer.net/pipermail/avahi-tickets/2014-December/001563.html
- http://mail.0pointer.net/pipermail/avahi-tickets/2014-December/001564.html


**Note**

The reason why I am seeking information is because there seems to be some confusion in userland about the correct usage of `IFA_LOCAL` vs. `IFA_ADDRESS` in IPv4 and IPv6. In my case I'm having problems with a ppp device; our code is querying IFA_ADDRESS to determine the interface address, which in case of a ppp device returns the peer address instead of the local address. The kernel headers contain the following hint, but that seems to be only half of the truth, since it does not seem to work in IPv6.

**/usr/include/linux/if_addr.h**

```
/*
 * Important comment:
 * IFA_ADDRESS is prefix address, rather than local interface address.
 * It makes no difference for normally configured broadcast interfaces,
 * but for point-to-point IFA_ADDRESS is DESTINATION address,
 * local address is supplied in IFA_LOCAL attribute.
```

As far as I understand your bugfixes, the correct solution is to query `IFA_LOCAL` first and use `IFA_ADDRESS` as fallback? Is there some official documentation which explains that and why this is the correct way of doing it, or is this the result of trial-and-error?

**Comment by lathiat** [comment]
Thanks for raising this issue, as I wonder if we are still getting this wrong sometimes.  Since we "reversed" the check twice.  It's great you're looking into the issue more comprehensively.

As you suggest the issue appears with interfaces that don't have a subnet but instead two endpoints.  PPP devices are the most common like this but you can also see it I think with tun devices.

The old tickets aren't online but they used to e-mail a list and I have an archive.  The original ticket is below.  It seems the original ticket was for an IPv6 case

From my reading of this bug and the attached Debian bug, the IPv6 interfaces were buggy at the time but I think it was not the entire story. As right now in the latest Avahi, both IPv6 and PtP interfaces work fine. The Debian bug suggests those were normal IPv6 interfaces and not PtP IPv6 interfaces

> #333: No longer works for interfaces with an IPv6 address
> --------------------+-------------------------------------------------------
>  Reporter:  mbiebl  |       Owner:  lennart   
>      Type:  defect  |      Status:  new       
> Milestone:          |   Component:  avahi-core
>  Keywords:          |  
> --------------------+-------------------------------------------------------
>  Starting with 0.6.29, avahi no longer registers an address record for
>  interfaces with an assigned IPv6 address.
> 
>  The faulty commit is 28948f72a2689afe0ba8a31fd396a0c7faf7c213
> 
> 
>  The bug was first reported via the Debian BTS and the complete log is
>  available at:
>  Bug-Debian: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=617948

**Comment by lathiat** [comment]
To clarify, the most recent bug fix I committed was to fix PPP interfaces specifically (side note: avahi does not activate on these interfaces by default, if you wanted to test avahi itself in this situation, you have to enable the allow-pointopoint option)

I believe the original fix for IPv6 interfaces fixed those but reversed the IFA_LOCAL/IFA_ADDRESS check at the same time.

**Comment by mspncp** [comment]
> Thanks for raising this issue, as I wonder if we are still getting this wrong sometimes. Since we "reversed" the check twice. It's great you're looking into the issue more comprehensively.

Thanks in return for your detailed reply. Unfortunately I won't be able to do any checks with the avahi daemon itself, because I'm not familiar with using it. But I did some more research which I would like to share.

I believe now that you are doing the right thing: First, [parse all attributes](https://github.com/lathiat/avahi/blob/v0.7/avahi-core/iface-linux.c#L203) and then [check for the presence](https://github.com/lathiat/avahi/blob/v0.7/avahi-core/iface-linux.c#L242) of IFA_LOCAL and IFA_ADDRESS. There are two cases

- If both are present, then IFA_LOCAL contains the local interface address and IFA_ADDRESS the peer address.
- If only IFA_ADDRESS is present  then it contains the local interface address.

It is indeed necessary to parse the entire attribute list before checking the two cases, because apparently for IPv4, rtnetlink will return first IFA_ADDRESS and then IFA_LOCAL, whereas for IPv6, it is the opposite way around. You can see this in the kernel source if you compare the implementations of `inet_fill_ifaddr()` and `inet6_fill_ifaddr`:

[/usr/src/linux/net/ipv4/devinet.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv4/devinet.c?h=v4.17#n1573):

```C
static int inet_fill_ifaddr(struct sk_buff *skb, struct in_ifaddr *ifa,
                u32 portid, u32 seq, int event, unsigned int flags)
{
    ...
    if ((ifa->ifa_address &&
         nla_put_be32(skb, IFA_ADDRESS, ifa->ifa_address)) ||       /* <=  first IFA_ADDRESS */
        (ifa->ifa_local &&
         nla_put_be32(skb, IFA_LOCAL, ifa->ifa_local)) ||           /* <=  then IFA_LOCAL */
                 ...
        goto nla_put_failure;

    ...
}
```

[/usr/src/linux/net/ipv6/addrconf.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv6/addrconf.c?h=v4.17#n4721):

```C
static int inet6_fill_ifaddr(struct sk_buff *skb, struct inet6_ifaddr *ifa,
                             u32 portid, u32 seq, int event, unsigned int flags)
{
        ...
        
        if (!ipv6_addr_any(&ifa->peer_addr)) {
                if (nla_put(skb, IFA_LOCAL, 16, &ifa->addr) < 0 ||       /* <=  first IFA_LOCAL */
                    nla_put(skb, IFA_ADDRESS, 16, &ifa->peer_addr) < 0)  /* <=  then  IFA_ADDRESS */
                        goto error;
        } else
                if (nla_put(skb, IFA_ADDRESS, 16, &ifa->addr) < 0)
                        goto error;

        ...
}
```

**Note:** Interestingly, glibc has a function [getifaddrs(3)](https://linux.die.net/man/3/getifaddrs) which obtains the interface addresses using rtnetlink. If you take a look at its implementation then you will see that it does not read all attributes before doing the check, which is probably the reason why their check is somewhat weird.


[~/src/glibc-2.26/sysdeps/unix/sysv/linux/ifaddrs.c](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/ifaddrs.c;h=32381f54e4e0e10c42c47aed0ebeb1df03bf19af;hb=23158b08a0908f381459f273a984c6fd328363cb#l313):

```C
static int
getifaddrs_internal (struct ifaddrs **ifap)
{
        ...

              while (RTA_OK (rta, rtasize))
                {
                  char *rta_data = RTA_DATA (rta);
                  size_t rta_payload = RTA_PAYLOAD (rta);

                  switch (rta->rta_type)
                    {
                    case IFA_ADDRESS:
                      {
                        struct sockaddr *sa;

                        if (ifas[ifa_index].ifa.ifa_addr != NULL)
                          {
                            /* In a point-to-poing network IFA_ADDRESS
                               contains the destination address, local
                               address is supplied in IFA_LOCAL attribute.
                               destination address and broadcast address
                               are stored in an union, so it doesn't matter
                               which name we use.  */
                            ifas[ifa_index].ifa.ifa_broadaddr
                              = &ifas[ifa_index].broadaddr.sa;
                            sa = &ifas[ifa_index].broadaddr.sa;
                          }
                        else
                          {
                            ifas[ifa_index].ifa.ifa_addr
                              = &ifas[ifa_index].addr.sa;
                            sa = &ifas[ifa_index].addr.sa;
                          }
                          
                        ...



                    case IFA_LOCAL:
                      if (ifas[ifa_index].ifa.ifa_addr != NULL)
                        {
                          /* If ifa_addr is set and we get IFA_LOCAL,
                             assume we have a point-to-point network.
                             Move address to correct field.  */
                          ifas[ifa_index].broadaddr = ifas[ifa_index].addr;
                          ifas[ifa_index].ifa.ifa_broadaddr
                            = &ifas[ifa_index].broadaddr.sa;
                          memset (&ifas[ifa_index].addr, '\0',
                                  sizeof (ifas[ifa_index].addr));
                        }
                        
                      ...

                  rta = RTA_NEXT (rta, rtasize);
                }
```

**Comment by mspncp** [comment]
**P.S:** While writing the previous post, I noticed that there is another `iface-linux.c` in `avahi-autoipd` which treats `IFA_LOCAL` and `IFA_ADDRESS` differently. You should have a closer look [at this location](https://github.com/lathiat/avahi/blob/v0.7/avahi-autoipd/iface-linux.c#L148), it might be possible that the implementation is incorrect.

**Comment by mspncp** [comment]
**P.P.S:** When I read that avahi 0.7, was called the "Adieu" release, my first thought was "Oh, was avahi discontinued?". Only after that it came to my mind that it's probably a pun on Apple's "Bonjour" ;-).

**Comment by mspncp** [comment]
Hi Trent,

thanks again for taking the time to answer my questions. I updated the issue's title to reflect the contents of this thread more properly. Feel free to close this issue and/or let me know whether I should open a separate issue for https://github.com/lathiat/avahi/issues/187#issuecomment-399929466.

Regards, Matthias
